arbor/include/arbor/morph/mcable_map.hpp

```diff
@@ -0,0 +1,121 @@
+#pragma once
+
+// Represent an mcable_list where each cable has an associated value.
+//
+// Values may be mutated via operator[]; otherwise the only mutating operations
+// are insert, emplace, and clear.
+
+#include <algorithm>
+#include <vector>
+
+#include <arbor/assert.hpp>
+#include <arbor/morph/primitives.hpp>
+#include <arbor/util/optional.hpp>
+
+namespace arb {
+
+template <typename T>
+struct mcable_map {
+    // Forward subset of vector interface:
+
+    using value_type = std::pair<mcable, T>;
+    using size_type = std::size_t;
+    using difference_type = std::ptrdiff_t;
+
+    using const_iterator = typename std::vector<value_type>::const_iterator;
+    using iterator = const_iterator;
+
+    using const_reverse_iterator = typename std::vector<value_type>::const_reverse_iterator;
+    using reverse_iterator = const_reverse_iterator;
+
+    const_iterator begin() const noexcept { return elements_.cbegin(); }
+    const_iterator cbegin() const noexcept { return elements_.cbegin(); }
+
+    const_reverse_iterator rbegin() const noexcept { return elements_.crbegin(); }
+    const_reverse_iterator crbegin() const noexcept { return elements_.crbegin(); }
+
+    const_iterator end() const noexcept { return elements_.cend(); }
+    const_iterator cend() const noexcept { return elements_.cend(); }
+
+    const_reverse_iterator rend() const noexcept { return elements_.crend(); }
+    const_reverse_iterator crend() const noexcept { return elements_.crend(); }
+
+    const value_type& operator[](size_type i) const { return elements_[i]; }
+
+    auto front() const { return *begin(); }
+    auto back() const { return *rbegin(); }
+
+    std::size_t size() const noexcept { return elements_.size(); }
+    bool empty() const noexcept { return !size(); }
+
+    void clear() { elements_.clear(); }
+
+    // mcable_map-specific operations:
+
+    // Insertion is successful iff c intersects with cables in at most two discrete points.
+    // insert() and emplace() return true on success.
+
+    bool insert(const mcable& c, T value) {
```

Comment:
I think the following is a bit easier to follow:
```C++
if (auto opt_it = insertion_point(c)) {
    elements_.insert(*opt_it, value_type{c, std::move(value)});
    assert_invariants();
    return true;
}
return false;
```
If you don't agree, fine, but there is no reason to use comma operator on the `return` statement!

Reply:
I prefer early-exit style code — for me, at least, it gets the exceptional case stuff out of the way, and then the 'main' part of the procedure can go on without being in some sub-block.

The comma operator is perfectly innocent: it doesn't need to have a speical reason to be used :D
But I'll pop the assertion on a separate line.